### PR TITLE
Fixed #10827 -- Ensured ContentTypes are created before permission creation

### PR DESCRIPTION
**Pull request by claudep**
https://code.djangoproject.com/ticket/10827

### CODE REVIEW

File: django/contrib/auth/management/__init__.py [hunk 1]
```diff
@@ -37,6 +37,11 @@ def create_permissions(app_config, verbosity=2, interactive=True, using=DEFAULT_
     if not app_config.models_module:
         return
 
+    # It's important to ensure here that create_contenttypes has been called for
+    # this app (we don't control post_migrate signal receiver's ordering).
+    from django.contrib.contenttypes.management import create_contenttypes
```
Comment (charettes) [inline django/contrib/auth/management/__init__.py]: Does it break anything if this is moved to the top? I would assume this module can safely depend on the later?

File: tests/auth_tests/test_management.py [hunk 2]
```diff
@@ -809,3 +809,14 @@ def test_unavailable_models(self):
...
+    def test_create_permissions_check_contenttypes_created(self):
+        """
+        `post_migrate` handler ordering is not guaranteed, so we simulate a
+        case here where create_permissions is called before create_contenttypes.
+        """
+        # Warm the manager cache.
+        ContentType.objects.get_for_model(Group)
+        # Apply a deletion as if e.g. a database 'flush' had been executed.
+        ContentType.objects.filter(app_label='auth', model='group').delete()
+        create_permissions(apps.get_app_config('auth'), interactive=False, verbosity=0)
```
[the remark this answers is not in the document]
Comment (claudep) [reply tests/auth_tests/test_management.py]: Yes, I was bitten by this today, and thought it is enough time lost by fellow developers!

File: django/contrib/auth/management/__init__.py [hunk 3]
```diff
@@ -37,6 +38,10 @@ def create_permissions(app_config, verbosity=2, interactive=True, using=DEFAULT_
     if not app_config.models_module:
         return
 
+    # It's important to ensure here that create_contenttypes has been called for
+    # this app (we don't control post_migrate signal receiver's ordering).
+    create_contenttypes(app_config, verbosity=verbosity, interactive=interactive, using=using, apps=apps, **kwargs)
+
```
Comment (charettes) [inline django/contrib/auth/management/__init__.py]: On thing to keep in mind is that this is going to be introducing double work.

**Comment by claudep** [comment]
Sure, this reminds us about the wider discussion of INSTALLED_APPS ordering. I agree on discussing this elsewhere, because this bug has waited for too long now.

**Pull request merged**